feat: CRUD Students / feat: Teachers & Users (pull request incompleta - falta Principals)

LucasDLabo opened the pull request:
### Student
- Se realizó el CRUD de Student con las correspondientes validaciones y mensajes de error.
- Se añadió la relación con la tabla "city" y "user" para su correcto funcionamiento.
- Se cambió el atributo authorize() dentro de la request a "true" para su correcto funcionamiento.
- Se tuvo que cambiar el update para que coloque null en user_id cuando se elimine el estudiante, para liberar el id de su user_id asignado.
- Se cambio la columna "user_id" para que pueda ser nullable() para su correcto funcionamiento.
- Se cambio el código de la factory para la tabla "students" para que solo se generen campos únicos en "user_id".
- Se agregó una etiqueta (Actual) para identificar cuál es la ciudad actual que tiene el estudiante a la hora de editar.
- Se agregó una función para mantener la ciudad y usuario elegido cuando se edita un estudiante.
- Se agregó una función para ordenar las ciudades y usuarios de manera alfabética en la lista desplegable del Create y Edit.

### Users
- Se debió crear más users con las factories para tener usuarios libres (de 10 a 40)
- Se debió de ordenar de manera alfabética la lista de todos los usuarios por su cantidad.

### Teacher
- Se tuvo que cambiar el update para que coloque null en user_id cuando se elimine el profesor, para liberar el id de su user_id asignado.
- Se agregó una etiqueta "(Actual)" para identificar cuál es la ciudad actual que tiene el estudiante a la hora de editar.
- Se agregó una función para mantener la ciudad y usuario elegido cuando se edita un profesor.
- Se agregó una función para ordenar las ciudades y usuarios de manera alfabética en la lista desplegable del Create y Edit.

LucasDLabo committed:
feat: CRUD Students / feat: Teachers & Users

Student
- Se realizó el CRUD de Student con las correspondientes validaciones y mensajes de error.
- Se añadió la relación con la tabla "city" y "user" para su correcto funcionamiento.
- Se cambió el atributo authorize() dentro de la request a "true" para su correcto funcionamiento.
- Se tuvo que cambiar el update para que coloque null en user_id cuando se elimine el estudiante, para liberar el id de su user_id asignado.
- Se cambio la columna "user_id" para que pueda ser nullable() para su correcto funcionamiento.
- Se cambio el código de la factory para la tabla "students" para que solo se generen campos únicos en "user_id".
- Se agregó una etiqueta (Actual) para identificar cuál es la ciudad actual que tiene el estudiante a la hora de editar.
- Se agregó una función para mantener la ciudad y usuario elegido cuando se edita un estudiante.
- Se agregó una función para ordenar las ciudades y usuarios de manera alfabética en la lista desplegable del Create y Edit.

Users
- Se debió crear más users con las factories para tener usuarios libres (de 10 a 40)
- Se debió de ordenar de manera alfabética la lista de todos los usuarios por su cantidad.

Teacher
- Se tuvo que cambiar el update para que coloque null en user_id cuando se elimine el profesor, para liberar el id de su user_id asignado.
- Se agregó una etiqueta "(Actual)" para identificar cuál es la ciudad actual que tiene el estudiante a la hora de editar.
- Se agregó una función para mantener la ciudad y usuario elegido cuando se edita un profesor.
- Se agregó una función para ordenar las ciudades y usuarios de manera alfabética en la lista desplegable del Create y Edit.

MaksiConGit reviewed (changes requested):
- [x] Algunos registros de alumnos se generan con números de teléfono con más de 14 caracteres, haciendo que cuando lo editás y guardás sin modificar nada, te sale la validación del teléfono.
- [x] Al crear un usuario, se selecciona la ciudad, lo ideal sería que primero seleccione el país, que después pueda seleccionar las provincias de ese país, y que al final pueda seleccionar las ciudades de esa provincia.
- [x] City debería tener relación con su Province, y Province con su Country. Con esto podrías lograr el punto anterior.
- [x] Por lo que entiendo, $studentsThatHasUser contiene únicamente el user_id de los estudiantes que sí tienen cuenta, no? Si es así, es medio confuso el nombre
- [x] Quitar el user_id del estudiante y profesor al borrarlo es un error, ya que la cuenta le pertenece a ese usuario, y no debería pertenecer a nadie más. Por como está, al eliminar a, por ejemplo, Pepito 33, podrías asignar su cuenta a Selenita 24, y Selenita vería todos los datos personales de Pepito.
- [x] En la route está agregado el model Student
- [x] En el show faltan los siguiente datos:
    - [x] Asistencia
    - [x] Notas
    - [x] Cursos
    - [x] Carreras
    - [x] Institución
    - [x] Provincia
    - [x] País
- [x] Cambiaste el edit del CRUD de Teacher, es mala práctica hacer pequeños cambios en sectores donde no estás programando. Quitá el cambio, y si lo querés hacer, hacé una pull request aparte.

LucasDLabo committed:
fix: Student v1.1

Se volvió a los cambios anteriores de la PR de teachers (edit de teacher y controller).
Se cambio la factory de student para que genere números telefónicos correctos.
Se elimino un linkeo necesario del model Students en Routes.
Se elimino la funcionalidad de liberar un id de user_id al eliminar un estudiante.

LucasDLabo committed:
feat: Select Dependiente de City / Installed Livewire

 - Se implementó Livewire
 - Se creo la funcion de select dependiente que permite ir mostrando a tiempo real que ciudades y provincias pertenece a cada pais. Además el componente Livewire permite recibir un usuario ya creado con su correspondiente ciudad y mostrarlo/modificarlo.

LucasDLabo committed:
refactor: Nombre de variables más entendibles

 - Se cambio el nombre de 2 variables para que se entienda que contienen

LucasDLabo committed:
feat: Más informacion en el show de Student

 - Ahora se muestran los campos de Province y Country dentro del Student.
 - Se cambio la migration de Users para que no se generen con instituciones borradas.
 - Ahora se muestra los cursos del Student organizado por carreras.

LucasDLabo committed:
feat: Mostrar Courses y Exams de Students

- Se implementó una lista de los cursos correspondientes de los estudiantes agrupados por carrera.
- Dentro de los cursos de los estudiantes, se implementó una lista de exámenes correspondiente al curso y se muestra información de la nota del exámen si el estudiante lo tomó.

LucasDLabo committed:
feat: Mostrar Ausencias de los alumnos por cursos

 - Se refactoriza la Tabla AttendanceRecord para que tenga una columna de course_student_id y poder organizar las asistencias entre cursos
 - Ahora se muestra las inasistencias de los alumnos que tuvieron en sus cursos.

LucasDLabo committed:
feat: Dependant Select for Teacher

 - Se agregó el select dependiente de las ciudades en el Modelo Teacher en su Create y Edit

LucasDLabo committed:
feat: Más información en el show de Teachers

 - Ahora se muestra el país, provincia y ciudad a la que pertenece y además de la institutición.

LucasDLabo committed:
feat: CRUD de Principals

 - Se realizó el CRUD de Principals con las correspondientes vistas, validaciones y mensajes de error.

LucasDLabo committed:
refactor: Agrupar courses por carreras.

 - Se simplificó el código para agrupar los cursos de un student por careers.

LucasDLabo committed:
refactor: Mostrar detalles del curso

 - Se refactorizó el código para mostrar los exámenes e inasistencias de un student en sus cursos específicos.
